sources/packages/backend/apps/workers/src/services/disbursement-schedule/disbursement-schedule.service.ts

```diff
@@ -0,0 +1,179 @@
+      },
+    });
+
+    if (!firstDisbursement) {
+      throw new CustomNamedError(
+        "Disbursement not found or one of its associations is missing.",
+        DISBURSEMENT_NOT_FOUND,
+      );
+    }
+
+    if (firstDisbursement.msfaaNumber?.id) {
+      throw new CustomNamedError(
+        "MSFAA number is already associated.",
+        APPLICATION_MSFAA_ALREADY_ASSOCIATED,
+      );
+    }
+
+    let msfaaNumberId: number;
+
+    const studentId =
+      firstDisbursement.studentAssessment.application.student.id;
+    const offeringIntensity =
+      firstDisbursement.studentAssessment.offering.offeringIntensity;
+    const applicationId = firstDisbursement.studentAssessment.application.id;
+
+    // Checks if there is an MSFAA that could be considered valid.
+    const existingValidMSFAANumber =
+      await this.msfaaNumberService.getCurrentValidMSFAANumber(
+        studentId,
+        offeringIntensity,
+      );
+    if (existingValidMSFAANumber) {
+      // Reuse the MSFAA that is still valid and avoid creating a new one.
+      msfaaNumberId = existingValidMSFAANumber.id;
+    } else {
+      // Get previously completed and signed disbursement of an application for the student
+      // to determine if an existing MSFAA is still valid.
+      const previousSignedDisbursement =
+        await this.getPreviouslySignedDisbursement(
+          studentId,
+          offeringIntensity,
+        );
+
+      let hasValidMSFAANumber = false;
+      if (previousSignedDisbursement) {
+        // checks if the MSFAA number is still valid.
+        // If the study period end date of the previously signed MSFAA is less than 2 years
+        // when compared to current study period start date, then MSFAA is considered to be valid.
+        hasValidMSFAANumber = this.msfaaNumberService.isMSFAANumberValid(
+          // Previously signed and completed application offering end date in considered the start date.
+          new Date(
+            previousSignedDisbursement.studentAssessment.offering.studyEndDate,
+          ),
+          // Start date of the offering of the current application is considered the end date.
+          new Date(firstDisbursement.studentAssessment.offering.studyStartDate),
+        );
+      }
+
+      if (hasValidMSFAANumber) {
+        // Reuse the MSFAA number.
+        msfaaNumberId = previousSignedDisbursement.msfaaNumber.id;
+      } else {
+        // Create a new MSFAA number in case the previous one is no longer valid.
+        const newMSFAANumber = await this.msfaaNumberService.createMSFAANumber(
+          studentId,
+          applicationId,
+          offeringIntensity,
+        );
+        msfaaNumberId = newMSFAANumber.id;
+      }
+    }
+    const msfaaNumber = { id: msfaaNumberId } as MSFAANumber;
+    firstDisbursement.msfaaNumber = msfaaNumber;
+    const disbursementsToUpdate = [firstDisbursement];
+
+    // If the assessment has second disbursement, associate MSFAA number
```

Comment:
As a sugestion to make the code from line 124 to 134 simple, the beginning of the method can be declared as:
```ts
const disbursements = await this.repo.find({.....
const [firstDisbursement, secondDisbursement] = disbursements ;
```
then the last part of the method wold be
```ts
disbursements.foreach(disbursement => disbursement.msfaaNumber = msfaaNumber);
this.repo.save(disbursements);
```